Remove 'session_state' from tokens

graziang opened the pull request:
The `sid` claim is now mapped via the new `SessionIdMapper` protocol mapper added by default to `basic` client scope introduced by #28243.
Before this the `IDToken.getSessionId()` method always returned the `session_state` value, now the mapper reads the value directly from the user session.
Not sure if it's better to completely remove by default the `sid` claim from access token or id token, more details about the claim specification in the latest issue comments.
@mposolda WDYT?

Closes #27624

graziang committed:
Remove session_state from tokens

Closes #27624

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

mposolda commented:
@graziang Thanks for the PR and for checking the specifications!

It is a long ago and don't remember the exact history, but I think that few years ago, the original name of the claim in the frontchannel/backchannel logout specifications was `session_state`, but then they changed it to be shorter name `sid`. The session management specification section you pointed https://openid.net/specs/openid-connect-session-1_0.html#CreatingUpdatingSessions kept the `session_state`, but note that it mentions it as a name of the parameter in the authentication response (not the name of the claim in the ID token). That is probably the history behind the comment in our code `session_state is deprecated, sid should be used instead` :-)

So I think the better option might be to keep the `sid` claim in our IDToken (and access tokens), but remove `session_state` claim from tokens by default. WDYT?

I suppose we can do something like this in the `IDToken` class:

```
@JsonProperty(SESSION_ID)
protected String sessionState;
```

and we won't need `sessionId` field on that class. This means that all the places in our code, which calls `idToken.getSessionState()` will still work fine.

The `session_state` claim could then be added by the protocol mapper. But I wonder if we really need to add this protocol mapper to the `basic` client scope by default? As if I understand correctly, we don't need `session_state` claim to be present by default in any of the tokens (ID Token, Access Token, Refresh Token or Lightweight access token). As it seems there is not any specification, which really supports `session_state` as token claim? People can still use the protocol mapper just if they explicitly want to add the `session_state` claim, which we can document in the migration guide. WDYT?

graziang commented:
@mposolda Thanks for the review and the clarification about history of `session_state` and `sid`, now it makes more sense to me.

As we discussed, I added the `sessionId` claim in IDToken and removed the `sessionState` field while keeping `getSessionState()` method as deprecated.
`SessionIdMapper` has been replaced with `SessionStateMapper` and removed from the `basic` client scope.

mposolda commented on docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc:
```diff
@@ -77,4 +77,10 @@ This scope contains preconfigured protocol mappers for the following claims:
+
+The `session_state` claim which contains the same value as the `sid` claim is now removed from all tokens as it is not required according to the OpenID Connect Front-Channel Logout and OpenID Connect Back-Channel Logout specifications. The `session_state` claim remains present in the Access Token Response in accordance with OpenID Connect Session Management specification.
+
+A new `Session State (session_state)` mapper is also included and can be assigned to client scopes (for instance `basic` client scope) to revert to the old behavior.
```
Is it possible to also include a small notice that method `setSessionState()` is removed from `IDToken` class and `getSessionState()` is made deprecated? It is possible that some people might use this?

mposolda replied:
Thanks! I see there are test failures. I guess they are related to the fact that our `keycloak.js` adapter is relying on the `session_state` claim from the token, which causes javascript adapter to not behave correctly. Looks it is here: https://github.com/keycloak/keycloak/blob/main/js/libs/keycloak-js/src/keycloak.js#L1012 .

Is it possible to fix this in `keycloak.js` and also mention in the upgrading guide another minor note that people using old version of `keycloak.js` adapter in their applications must add the `SessionStateMapper` to their clients? We did similar already for the `nonce` parameter in the file `changes-25_0_0.adoc`, so maybe that section can be updated somehow to handle both `nonce` and `session_state` ?

minusdavid replied:
Thanks for sharing that bit of documentation. It looks like it wasn't adopted in the end? According to https://www.keycloak.org/2024/06/keycloak-2500-released and anecdotal experience, session_state isn't included in any claims (including the access token) these days.

mposolda reviewed (commented):
@graziang Thanks! Added one small note to the docs. Could you please check it?

Is it also possible to add some method for testing the `SessionStateMapper` ? (EG. adding the mapper and then check that `session_state` claim is present?) Thanks!

@andymunro Are you please able to review the documentation changes?

graziang commented:
> @mposolda Thanks for the review and the clarification about history of `session_state` and `sid`, now it makes more sense to me.
> 
> As we discussed, I added the `sessionId` claim in IDToken and removed the `sessionState` field while keeping `getSessionState()` method as deprecated. `SessionIdMapper` has been replaced with `SessionStateMapper` and removed from the `basic` client scope.

@mposolda Thanks! Tests for the `SessionNoteMapper` are already present in the `LightWeightAccessTokenTest` class, you can check line 600 for configuration  and `assertMapperClaims` method for check the mapped claims (line 400). Are tests here enough?

graziang commented:
> @graziang Thanks! Added one small note to the docs. Could you please check it?
> 
> Is it also possible to add some method for testing the `SessionStateMapper` ? (EG. adding the mapper and then check that `session_state` claim is present?) Thanks!
> 
> @andymunro Are you please able to review the documentation changes?

@mposolda Thanks! Tests for the SessionNoteMapper are already present in the LightWeightAccessTokenTest class, you can check line 600 for configuration and assertMapperClaims method for check the mapped claims (line 400). Are tests here enough?

mposolda commented:
> > @graziang Thanks! Added one small note to the docs. Could you please check it?
> > Is it also possible to add some method for testing the `SessionStateMapper` ? (EG. adding the mapper and then check that `session_state` claim is present?) Thanks!
> > @andymunro Are you please able to review the documentation changes?
> 
> @mposolda Thanks! Tests for the SessionNoteMapper are already present in the LightWeightAccessTokenTest class, you can check line 600 for configuration and assertMapperClaims method for check the mapped claims (line 400). Are tests here enough?

Yes, I completely overlooked it :-) Thanks for the clarification!

jonkoops commented on js/libs/keycloak-js/src/keycloak.js:
```diff
@@ -1009,7 +1009,7 @@ function Keycloak (config) {
             kc.token = token;
             kc.tokenParsed = jwtDecode(token);
-            kc.sessionId = kc.tokenParsed.session_state;
+            kc.sessionId = kc.tokenParsed.sid;
```
Was this `sid` present before? Users might use different versions of Keycloak, so if `sid` was introduced in this PR we should still fall back to `session_state` in case it doesn't exist.

graziang replied:
`sid` was already present as duplicate of `session_state`, `session_state` is now removed, so if an old adapter is used with a new version of Keycloak (which includes this PR) the `Session State` mapper must be added (documented it in the release changes).

jonkoops replied:
Ok, perfect, just wanted to do a sanity check.

graziang replied:
Thanks!

andymunro commented on docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc:
```diff
@@ -77,4 +77,14 @@ This scope contains preconfigured protocol mappers for the following claims:
 
-This helps to reduce even more the number of claims in a lightweight access token, but also gives the chance to configure claims that were always added automatically.
\ No newline at end of file
+This helps to reduce even more the number of claims in a lightweight access token, but also gives the chance to configure claims that were always added automatically.
```
```suggestion
This provides additional help to reduce the number of claims in a lightweight access token, but also gives the chance to configure claims that were always added automatically.
```

graziang replied:
Thanks, Done!

andymunro commented on docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc:
```diff
@@ -77,4 +77,14 @@ This scope contains preconfigured protocol mappers for the following claims:
+
+= Removed `session_state` claim
+
+The `session_state` claim which contains the same value as the `sid` claim is now removed from all tokens as it is not required according to the OpenID Connect Front-Channel Logout and OpenID Connect Back-Channel Logout specifications. The `session_state` claim remains present in the Access Token Response in accordance with OpenID Connect Session Management specification.
```
```suggestion
The `session_state` claim, which contains the same value as the `sid` claim, is now removed from all tokens as it is not required according to the OpenID Connect Front-Channel Logout and OpenID Connect Back-Channel Logout specifications. The `session_state` claim remains present in the Access Token Response in accordance with OpenID Connect Session Management specification.
```

graziang replied:
Thanks, Done!

andymunro commented on docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc:
```diff
@@ -77,4 +77,14 @@ This scope contains preconfigured protocol mappers for the following claims:
+
+The `session_state` claim which contains the same value as the `sid` claim is now removed from all tokens as it is not required according to the OpenID Connect Front-Channel Logout and OpenID Connect Back-Channel Logout specifications. The `session_state` claim remains present in the Access Token Response in accordance with OpenID Connect Session Management specification.
+
+Note that the `setSessionState()` method is also removed from `IDToken` class in favor of the `setSessionId()` method, and the `getSessionState()` method is now deprecated.
```
```suggestion
Note that the `setSessionState()` method is also removed from the `IDToken` class in favor of the `setSessionId()` method, and the `getSessionState()` method is now deprecated.
```

andymunro reviewed (commented):
A few suggestions,  @graziang

mposolda reviewed (approved):
@graziang @jonkoops @andymunro Thanks everyone for the review and updates!

minusdavid commented:
For folk coming here from elsewhere, you can re-add the "session_state" claim to your token using the "Session State (session_state)" mapper. 

You'll likely need to do this for libraries that support OpenID Connect Session Management like Drupal's Keycloak OpenID Connect module.
